niceman/distributions/debian.py

```diff
@@ -368,93 +365,212 @@ def _find_all_sources(self):
+                                             p.get("architecture")))
+            if not r:
+                r = cmd_results.get(p.get("name"))
+            if not r:
+                lgr.warning("Was unable to run dpkg -s for %s" %
+                            p.get("name"))
+                continue
+            p["architecture"] = r.get("Architecture")
+            p["version"] = r.get("Version")
+
+    @staticmethod
+    def create_lookup_from_apt_cache_show(cmd_results):
+        lookup_results = {}
+        for r in cmd_results:
+            lookup_results[r.get("Package")] = r
+            lookup_results["%s:%s" % (r.get("Package"),
+                                      r.get("architecture"))] = r
+        return lookup_results
+
+    def _get_pkgs_details_from_apt_cache_show(self, pkg_dicts):
+        # Convert package names to name:arch format
+        queries = []
+        for p in pkg_dicts:
+            # Use "dpkg -s pkg" to get the installed version and arch
+            architecture = p.get("architecture")
+            name = p.get("name")
+            version = p.get("version")
+            queries.append("%s=%s" % (name, version) if not architecture
+                           else "%s:%s=%s" % (name, architecture, version))
+
+        # Find out how many packages we can query at once
+        max_len = max([len(q) for q in queries])
+        num_queries = max((_MAX_LEN_CMDLINE - 13) // (max_len + 1), 1)
+
+        cmd_results = []
+        # Execute dpkg -s, parse, and collate the output
+        for subqueries in (queries[pos:pos + num_queries]
+                           for pos in range(0, len(queries), num_queries)):
+            try:
+                out, _ = self._session.execute_command(
+                    ['apt-cache', 'show'] + subqueries
+                )
+                out = utils.to_unicode(out, "utf-8")
+                cmd_results += parse_apt_cache_show_pkgs_output(out)
+            except CommandError as _:
+                raise  # Currently raise any exception we encounter
+
+        # Turn apt-cache show results into a lookup table by package name
+        cmd_results = self.create_lookup_from_apt_cache_show(cmd_results)
+
+        for p in pkg_dicts:
+            # find the version table in the parsed command results for the
+            # current package
+            r = cmd_results.get("%s:%s" % (p.get("name"),
+                                             p.get("architecture")))
+            if not r:
+                r = cmd_results.get(p.get("name"))
+            if not r:
+                lgr.warning("Was unable to run apt-cache show for %s" %
+                            p.get("name"))
+                continue
+            p["source_name"] = r.get("Source_name")
+            p["source_version"] = r.get("Source_version")
+            p["size"] = r.get("Size")
+            p["md5"] = r.get("MD5sum")
+            p["sha1"] = r.get("SHA1")
+            p["sha256"] = r.get("SHA256")
```

Comment:
if those should be defined, better imho to use `[]` not `get`, again to catch problems early. I would use `get` more sparingly -- really in the cases where I expect that there might be no that key.

And since I don't like duplication, I would have done it like (again -- just nitpicking, please ignore unless you like it too ;))
```python
for f in ('Source_name', 'Source_version', 'Size', ...):
   # only some fields needs custom renaming
   p[{'MD5sum': 'md5'}.get(f, f)] = r[f]
```
which might be worthwhile of a little helper `map_fields(from, to, mapping={})`

Reply:
I pushed the lowercase and md5 to the parsing support function, simplifying the code here.